File: code/commons/src/main/java/net/shrine/protocol/CRCRequestType.java
package net.shrine.protocol;

import scala.Option;

/**
 * Simple enum listing the types of CRC requests that Shrine knows how to
 * handle. The enum names are deliberately the same as the jaxb generated class
 * of the request.
 *
 * @author Justin Quan
 * @date Jun 14, 2010
 * @link http://cbmi.med.harvard.edu
 * @link http://chip.org
 * <p/>
 * NOTICE: This software comes with NO guarantees whatsoever and is
 * licensed as Lgpl Open Source
 * @link http://www.gnu.org/licenses/lgpl.html
 */
public enum CRCRequestType {
    GetPDOFromInputListRequestType("getPDO_fromInputList"),
    InstanceRequestType("CRC_QRY_getQueryResultInstanceList_fromQueryInstanceId"),
    MasterRequestType("CRC_QRY_getQueryInstanceList_fromQueryMasterId"),
    QueryDefinitionRequestType("CRC_QRY_runQueryInstance_fromQueryDefinition"),
    UserRequestType("CRC_QRY_getQueryMasterList_fromUserId"),
    SheriffRequestType(null),
    ResultListRequestType("CRC_QRY_getQueryResultInstanceList_fromQueryInstanceId"),
    ResultRequestType("CRC_QRY_getResultDocument_fromResultInstanceId"),
    MasterDeleteRequestType("CRC_QRY_deleteQueryMaster"),
    MasterRenameRequestType("CRC_QRY_renameQueryMaster"),
    GetRequestXml("CRC_QRY_getRequestXml_fromQueryMasterId");
    
    public final Option<String> i2b2RequestType;

    private CRCRequestType(final String i2b2RequestType)
    {
        this.i2b2RequestType = option(i2b2RequestType);
    }
    
    private static Option<String> option(final String s) {
        return Option.apply(s);
    }
}

File: code/commons/src/main/scala/net/shrine/protocol/CrcRequestType.scala
package net.shrine.protocol

import net.shrine.util.SEnum

/**
 * Simple enum listing the types of CRC requests that Shrine knows how to
 * handle. The enum names are deliberately the same as the jaxb generated class
 * of the request.
 *
 * @author Justin Quan
 * @author clint
 * @date Jun 14, 2010
 * @link http://cbmi.med.harvard.edu
 * @link http://chip.org
 * <p/>
 * NOTICE: This software comes with NO guarantees whatsoever and is
 * licensed as Lgpl Open Source
 * @link http://www.gnu.org/licenses/lgpl.html
 */
final class CrcRequestType private[CrcRequestType] (val name: String, val i2b2RequestType: String) extends CrcRequestType.Value

//TODO: Rename these, add unit test

object CrcRequestType extends SEnum[CrcRequestType] {
  val GetPDOFromInputListRequestType = new CrcRequestType("GetPDOFromInputListRequestType", "getPDO_fromInputList")
  
  val InstanceRequestType = new CrcRequestType("InstanceRequestType", "CRC_QRY_getQueryResultInstanceList_fromQueryInstanceId")
  
  val MasterRequestType = new CrcRequestType("MasterRequestType", "CRC_QRY_getQueryInstanceList_fromQueryMasterId")
  
  val QueryDefinitionRequestType = new CrcRequestType("QueryDefinitionRequestType", "CRC_QRY_runQueryInstance_fromQueryDefinition")
  
  val UserRequestType = new CrcRequestType("UserRequestType", "CRC_QRY_getQueryMasterList_fromUserId")
  
  val ResultRequestType = new CrcRequestType("ResultRequestType", "CRC_QRY_getResultDocument_fromResultInstanceId")
  
  val MasterDeleteRequestType = new CrcRequestType("MasterDeleteRequestType", "CRC_QRY_deleteQueryMaster")
  
  val MasterRenameRequestType = new CrcRequestType("MasterRenameRequestType", "CRC_QRY_renameQueryMaster")
  
  val GetRequestXml = new CrcRequestType("GetRequestXml", "CRC_QRY_getRequestXml_fromQueryMasterId")
}
